wpilibNewCommands/src/main/native/include/frc2/command/CommandPtr.h

```diff
@@ -0,0 +1,189 @@
+   *
+   * @param toRun the Runnable to run
+   * @param requirements the required subsystems
+   * @return the decorated command
+   */
+  [[nodiscard]] CommandPtr AndThen(
+      std::function<void()> toRun,
+      wpi::span<Subsystem* const> requirements = {}) &&;
+
+  /**
+   * Decorates this command with a runnable to run after the command finishes.
+   *
+   * @param toRun the Runnable to run
+   * @param requirements the required subsystems
+   * @return the decorated command
+   */
+  [[nodiscard]] CommandPtr AndThen(
+      std::function<void()> toRun,
+      std::initializer_list<Subsystem*> requirements) &&;
+
+  /**
+   * Decorates this command with a runnable to run before this command starts.
+   *
+   * @param toRun the Runnable to run
+   * @param requirements the required subsystems
+   * @return the decorated command
+   */
+  [[nodiscard]] CommandPtr BeforeStarting(
+      std::function<void()> toRun,
+      std::initializer_list<Subsystem*> requirements) &&;
+
+  /**
+   * Decorates this command with a runnable to run before this command starts.
+   *
+   * @param toRun the Runnable to run
+   * @param requirements the required subsystems
+   * @return the decorated command
+   */
+  [[nodiscard]] CommandPtr BeforeStarting(
+      std::function<void()> toRun,
+      wpi::span<Subsystem* const> requirements = {}) &&;
+
+  /**
+   * Decorates this command with a timeout.  If the specified timeout is
+   * exceeded before the command finishes normally, the command will be
+   * interrupted and un-scheduled.  Note that the timeout only applies to the
+   * command returned by this method; the calling command is not itself changed.
+   *
+   * @param duration the timeout duration
+   * @return the command with the timeout added
+   */
+  [[nodiscard]] CommandPtr WithTimeout(units::second_t duration) &&;
+
+  /**
+   * Decorates this command with an interrupt condition.  If the specified
+   * condition becomes true before the command finishes normally, the command
+   * will be interrupted and un-scheduled. Note that this only applies to the
+   * command returned by this method; the calling command is not itself changed.
+   *
+   * @param condition the interrupt condition
+   * @return the command with the interrupt condition added
+   */
+  [[nodiscard]] CommandPtr Until(std::function<bool()> condition) &&;
+
+  /**
+   * Decorates this command to only run if this condition is not met. If the
+   * command is already running and the condition changes to true, the command
+   * will not stop running. The requirements of this command will be kept for
+   * the new conditonal command.
+   *
+   * @param condition the condition that will prevent the command from running
+   * @return the decorated command
+   */
+  [[nodiscard]] CommandPtr Unless(std::function<bool()> condition) &&;
+
+  /**
+   * Get a raw pointer to the held command.
+   */
+  Command* operator*() const;
```

Comment:
`operator*()` of std::unique_ptr returns a reference to the held object, so I think the same should be done here by returning `Command&`.

Should the pointer aspect be leaned into more by defining `operator->()`? I think that just provides two ways to do the same thing and should be avoided. CommandPtr also behaves more like a value type than a pointer type (e.g., you use `.` to call member functions).

Reply:
CommandPtr's whole point is that it's passed by value moves.
As for `operator*` returning an lvalue reference, sure -- I just want to make sure folks don't do weird stuff like moving the underlying command out of the pointer, which would cause behavior I'm not sure of.
As for `operator->`, I don't see a reason: what would that operator return? Bc the whole rationale for CommandPtr as opposed to unique_ptr was so we could define the decorators on it.

Reply:
https://en.cppreference.com/w/cpp/memory/unique_ptr/operator*

Since pointer semantics aren't intended, I'd just remove CommandPtr::operator* entirely.

Reply:
That operator is there for internal implementation (ie, the scheduler uses `Command*`). It's not intended for extended usage by teams, but should still be available to them as an escape hatch.

Reply:
One of the examples uses it though, and `operator*` is conventionally not supposed to return a pointer. I'd prefer a getter over an operator overload.

Reply:
Oh right I totally forgot that GetAutoCommand in the templates/examples uses Command*, thus needing the underlying pointer.
It might be better to migrate them to use CommandPtr instead, but that would probably increase changes -- what do you think?

---
As for `get` vs `operator*` I totally agree.

Reply:
If it's mostly simple type renames, we can do the transition in this PR.

Reply:
On second thought, a primary difference between Command* and CommandPtr is nullability, which I'm not sure is trivially migrateable. I can add a null construction and emptiness checks in CommandPtr, but I'm not sure if that's a good idea.
I need to check what happens with a moved-from CommandPtr.